Add Reactive CORS support

sdeleuze opened the pull request:
This is a draft PR not ready yet to be merged, only here for review and discussion.

It implements CORS support by creating reactive version of Spring Web and Spring MVC CORS support classes:
- I chose to keep the same class names, but in `reactive` package.
- `CorsConfiguration` is not tied to Servlet API so we can reuse it
- CORS `HandlerMapping` integration is done at `AbstractHandlerMapping` level. Since we can't add dynamically filters CorsProcessor#processRequest()` is directly called.
- `AbstractUrlHandlerMapping` and `AbstractHandlerMethodMapping` have been slightly modified to call `AbstractHandlerMapping#processCorsRequest()`
- Only support for global CORS configuration has been implemented, `@CrossOrigin` support will come with another commit.

Could you make a quick high-level overview of this PR @rstoyanchev, before I go further?

rstoyanchev commented:
@sdeleuze some initial comments.

* I don't think the two cors packages (servlet and reactive) need to be co-located. I haven't checked for package cycles but the dependency from the `web.cors` on `web.server` is a bit bothersome. We did not have `web.server` before the reactive support. I'd be in favor of just having `web.cors.reactive` under `web.server.cors` where it fits better with the reactive support.
* For consistency both `CorsConfigurationSource` and `CorsProcessor` should take `ServerWebExchange`.
* The `Mono<Boolean>` return value on `CorsProcessor` is quite surprising. It seems to come down to the `writeWith(Mono.empty()` at the end (shouldn't that be `setComplete()` btw?) I would rather see the `CorsProcessor` only updating headers and returning a plain `boolean` leaving it to the caller to call `setComplete`.
* As a follow-up on the last point, shouldn't `AbstractHandlerMapping` be checking the return value from `processRequest` and returning a NoOpHandler if the CORS check failed?

rstoyanchev commented:
On second thought maybe `web.cors.reactive` is the way it should be with `web.server` being a low level package mostly depending on `web.http` and other packages under `web` (e.g. cors, bind, etc) in turn depending on `web.server`.

sdeleuze commented:
I have updated the commit based on your feedback. I also made some fixes that allows CORS support to work also on Reactor and RxNetty (see `Fix RxNetty/ReactorServerHttpRequest URI handling`commit).

I am now going to work on `@CrossOrigin` support.

sdeleuze committed:
Add Reactive CORS support

This is a port of Spring MVC CORS support for Spring Web Reactive:
 - CORS classes keep the same name but are in the
   web.cors.reactive package
 - CorsConfiguration is reused because not tied to Servlet API
 - CORS HandlerMapping integration is done at
   AbstractHandlerMapping level
 - AbstractUrlHandlerMapping and AbstractHandlerMethodMapping
   have been slightly modified to call
   AbstractHandlerMapping#processCorsRequest()
 - Both global CORS configuration + @CrossOrigin support have
   been implemented

Issue: SPR-14545

sdeleuze committed:
Add Reactive CORS unit and integration tests

Issue: SPR-14545

sdeleuze commented:
I think this PR is now ready to be reviewed before a merge on master, it now contains support for `@CrossOrigin`.

sdeleuze committed:
Introduce CorsUtils#isSameOrigin()

This commit modifies DefaultCorsProcessor to use a new
CorsUtils#isSameOrigin() method instead of WebUtils#isSameOrigin()
in order to avoid NoClassDefFoundError with non Servlet-based
engines.

Issue: SPR-14545

sdeleuze committed:
Restore Reactor @CrossOrigin tests

These tests are now green after the upgrade to Reactor
Netty 0.5.2.

Issue: SPR-14545

sdeleuze committed:
Improve CORS processing when scheme, host or port is null

This commit fix CORS processing when scheme is null, and
make additional assertions to check that host and port are
defined.

Issue: SPR-14545

rstoyanchev commented:
I've processed the PR applying some improvements such as re-ordering of fields and methods, test improvements, and I've also created a `getHandler` method at the `AbstractHandlerMapping` level that drives the CORS checks.
